### ISSUE TITLE
Evaluation in 1 Out-of-Domain Point for Verifier

### ISSUE THREAD

**Issue by aszepieniec** [TritonVM/triton-vm#133]
During last Friday's call with @bobbinth and @grtje an architectural difference between Miden and Triton came to light. The Miden verifier evaluates the AIR constraints in one out-of-domain point, whereas the Triton verifier evaluates them in $t$ points where $t$ is the number of colinearity checks for FRI.

Here is how Triton VM works in more detail:

 - The prover commits to the base (main?) columns as well as the randomizer, by zipping the codewords over the FRI domain and putting them into a Merkle tree.
 - The verifier supplies random challenges via Fiat-Shamir.
 - The prover extends the tables by computing the extension (auxiliary) columns.
 - The prover evaluates the AIR constraints and divides out the respective zerofiers, a process that results in the quotients.
 - The prover commits to the extension columns by zipping the codewords over the FRI domain and putting them into a Merkle tree. The quotients are not committed to.
 - The prover combines the randomizer, the base columns, the extension columns, and quotients into one nonlinear combination with random weights.
 - The prover and verifier run FRI on this nonlinear combination codeword to verify that it has low degree.
 - The verifier verifies that the nonlinear combination codeword is correctly formed in all the points probed by FRI. In these points it obtains the randomizer and base column values by authenticating them against the first Merkle root; it obtains the extension column values by authenticating them against the second Merkle root; and it obtains the quotient values by evaluating the AIR constraints.

Here is how *I think* Miden VM works. Or at least, how Triton VM could work. I'm sure there are more architectural differences, but the one that's relevant to this discussion is captured here:

 - The prover commits to the main (base) columns as well as the randomizer by zipping the codewords over the FRI domain and putting them into a Merkle tree.
 - The verifier supplies random challenges via Fiat-Shamir.
 - The prover extends the tables by computing the auxiliary (extension) columns.
 - The prover evaluates the AIR contraints and divides out the respective zerofiers. This process also results in quotients.
 - Using random weights supplied by the verifier the many quotients can be compressed into one weighted sum.
 - The prover commits to the extension columns **and** the quotient-sum, by zipping the codewords over the FRI domain and putting them into a Merkle tree.
 - The verifier samples a single out-of-domain point uniformly at random from the large field: $z \xleftarrow{\$} \mathbb{F}_p$.
 - The prover sends the values of the randomizer, main (base) polynomials, and auxiliary (extension) polynomials when evaluated in $z$ and, for all polynomials involved in transition constraints with distance $\omega$, $\omega \cdot z$.
 - The prover and verifier separately calculate the matching quotient-sum point, *i.e.*, evaluated in $z$.
 - All {main, auxiliary, quotient-sum} polynomials have an associated interpolant through revealed points and an associated zerofier; therefore every one of these polynomials corresponds to a new quotient obtained by subtracting the interpolant and dividing out the zerofier. These second-order quotients are combined into a second random combination using another batch of weights supplied by the verifier.
 - The prover and verifier run FRI to verify that this second random combination has low degree.
 - The verifier checks that this second random combination is correctly formed in each point queried by FRI by verifying the inner product between weights and polynomial values. It obtains the polynomial values from main and {auxiliary, quotient-sum} trees, followed by subtracting the interpolant and dividing out the zerofier.

PROs of the new approach:

 - Verifier is faster: current it has to evaluate all AIR constraints for all points probed by FRI; whereas with this impovement it has to evaluate all AIR constraints once, and do a relatively small inprocessing operation to get the points probed by FRI.

CONs of the new approach:

 - Proof is (slightly) larger: it needs to supply the evaluations in $z$ and the leafs of the extension-tree or {auxiliary, quotient-sum}-tree are larger by one large field element.
 - Prover is slower: it needs to evaluate a bunch of polynomials in $z$, calculate another weighted sum, and do an inprocessing step on a bunch of entire codewords.

**Comment by bobbinth** [comment]
> CONs of the new approach:
> 
> * Proof is (slightly) larger: it needs to supply the evaluations in z and the leafs of the extension-tree or {auxiliary, quotient-sum}-tree are larger by one large field element.
> * Prover is slower: it needs to evaluate a bunch of polynomials in z, calculate another weighted sum, and do an inprocessing step on a bunch of entire codewords.

I agree with the second CON, but I'm not sure about the first one: I think the proofs are actually smaller. Specifically, the the first approach the verifier needs to evaluate constraints at all queried points. This means that the verifier needs to have access to both `current` and `next` rows at the queried point.

In the second approach, the verifier does need to have the `current` and `next` rows at $z$, but for all other queried points they only need to have a single row. So, if we are doing more than 5 queries or so, the proof size should be smaller in the second approach.

**Comment by aszepieniec** [comment]
Agreed.

PRO 2: Number of randomizers can be smaller. With the current approach, the number of trace randomizers must be at least $4t$. One doubling comes from FRI, and one doubling comes from needing to open `current` and `next` rows. With the new approach, we need only $2t+2$ trace randomizers. One doubling still comes from FRI, but there is no second doubling. The $+2$ comes from releasing the evaluations in $z$ and $\omega \cdot z$.

**Comment by jan-ferdinand** [comment]
Fixed in 9606441383579bee135540759f608d50ff3508d4.
